File: CN/stop_and_wait/receiver.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <cstdlib>
#pragma comment(lib, "ws2_32.lib")

using namespace std;

int main() {
    WSADATA wsa;
    WSAStartup(MAKEWORD(2,2), &wsa);

    SOCKET sockfd;
    char buffer[1024];
    sockaddr_in receiverAddr{}, senderAddr{};
    int addr_len = sizeof(senderAddr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    receiverAddr.sin_family = AF_INET;
    receiverAddr.sin_port = htons(5001);
    receiverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    bind(sockfd, (sockaddr*)&receiverAddr, sizeof(receiverAddr));

    int expected = 0;

    cout << "RECEIVER STARTED\n\n";

    while (true) {
        memset(buffer, 0, sizeof(buffer));

        recvfrom(sockfd, buffer, sizeof(buffer), 0,
                 (sockaddr*)&senderAddr, &addr_len);

        cout << "RECEIVER: Frame arrived -> " << buffer << endl;

        int seq = buffer[6] - '0';

        // 1. Simulate frame loss
        if (rand() % 100 < 20) {
            cout << "RECEIVER: Frame lost\n\n";
            continue;
        }

        // 2. Simulate corruption
        bool corrupted = (rand() % 100 < 20);
        if (corrupted) {
            cout << "RECEIVER: Frame corrupted\n";

            char nak[10];
            sprintf(nak, "NAK:%d", seq);

            sendto(sockfd, nak, strlen(nak), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: NAK sent -> " << nak << "\n\n";
            continue;
        }

        if (seq == expected) {
            cout << "RECEIVER: Correct frame received\n";

            // 3. Simulate delayed ACK
            if (rand() % 100 < 20) {
                cout << "RECEIVER: Delaying ACK...\n";
                Sleep(3000);
            }

            char ack[10];
            sprintf(ack, "ACK:%d", seq);

            // 4. Simulate ACK loss
            if (rand() % 100 < 20) {
                cout << "RECEIVER: ACK lost\n\n";
                continue;
            }

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: ACK sent -> " << ack << endl;

            expected = 1 - expected;
        } else {
            cout << "RECEIVER: Duplicate frame\n";

            char ack[10];
            sprintf(ack, "ACK:%d", 1 - expected);

            sendto(sockfd, ack, strlen(ack), 0,
                   (sockaddr*)&senderAddr, addr_len);

            cout << "RECEIVER: Re-sent ACK -> " << ack << endl;
        }

        cout << endl;
    }

    closesocket(sockfd);
    WSACleanup();
    return 0;
}

File: PRAC_EXAM/CN/4_DLL/stop_and_wait/receiver.py
import socket
import random
import time

HOST = "127.0.0.1"
PORT = 5000

LOSS_RATE = 0.2
CORRUPT_RATE = 0.2
ACK_LOSS_RATE = 0.2

server = socket.socket()
server.bind((HOST, PORT))
server.listen(1)

print("RECEIVER STARTED\n")

conn, addr = server.accept()
print("Connected:", addr)

expected = 0

while True:

    data = conn.recv(1024)

    if not data:
        break

    frame = data.decode()
    print("Received:", frame)

    seq = int(frame.split(":")[1])

    # Simulate frame loss
    if random.random() < LOSS_RATE:
        print("Frame lost\n")
        continue

    # Simulate corruption
    if random.random() < CORRUPT_RATE:
        nak = f"NAK:{seq}"
        conn.send(nak.encode())
        print("Frame corrupted")
        print("Sent:", nak, "\n")
        continue

    # Correct frame
    if seq == expected:

        print("Correct frame received")

        # Simulate delayed ACK
        if random.random() < 0.2:
            print("Delaying ACK...")
            time.sleep(3)

        ack = f"ACK:{seq}"

        # Simulate ACK loss
        if random.random() < ACK_LOSS_RATE:
            print("ACK lost\n")
            continue

        conn.send(ack.encode())
        print("Sent:", ack, "\n")

        expected = 1 - expected

    # Duplicate frame
    else:
        ack = f"ACK:{1 - expected}"
        conn.send(ack.encode())

        print("Duplicate frame")
        print("Re-sent:", ack, "\n")

conn.close()
server.close()